gc/policy/policy.go

```diff
@@ -0,0 +1,217 @@
+				"MutationThreshold": fmt.Sprint(m.mutationThreshold),
+				"ScheduleDelay":     fmt.Sprint(m.scheduleDelay),
+			}
+
+			go m.run(ic.Context)
+
+			return m, nil
+		},
+	})
+}
+
+type mutationEvent struct {
+	ts    time.Time
+	dirty bool
+}
+
+type collector interface {
+	RegisterMutationCallback(func(bool))
+	GarbageCollect(context.Context) error
+}
+
+type gcPolicyManager struct {
+	c collector
+
+	eventC chan mutationEvent
+
+	pauseThreshold    float64
+	onDelete          bool
+	mutationThreshold int
+	scheduleDelay     time.Duration
+	startupDelay      time.Duration
+}
+
+func newPolicyManager(c collector, cfg *config) *gcPolicyManager {
+	eventC := make(chan mutationEvent)
+
+	m := &gcPolicyManager{
+		c:                 c,
+		eventC:            eventC,
+		pauseThreshold:    cfg.PauseThreshold,
+		onDelete:          cfg.ScheduleOnDelete,
+		mutationThreshold: cfg.MutationThreshold,
+		scheduleDelay:     time.Duration(cfg.ScheduleDelayMs) * time.Millisecond,
+		startupDelay:      time.Duration(cfg.StartupDelayMs) * time.Millisecond,
+	}
+
+	if m.pauseThreshold < 0.0 {
+		m.pauseThreshold = 0.0
+	}
+	if m.pauseThreshold > 0.5 {
+		m.pauseThreshold = 0.5
+	}
+	if m.mutationThreshold < 0 {
+		m.mutationThreshold = 0
+	}
+	if m.scheduleDelay < 0 {
+		m.scheduleDelay = 0
+	}
+	if m.startupDelay < 0 {
+		m.startupDelay = 0
+	}
+
+	c.RegisterMutationCallback(m.mutationCallback)
+
+	return m
+}
+
+func (m *gcPolicyManager) mutationCallback(dirty bool) {
+	e := mutationEvent{
+		ts:    time.Now(),
+		dirty: dirty,
+	}
+	go func() {
+		m.eventC <- e
+	}()
+}
+
+func schedule(d time.Duration) (s <-chan time.Time, t *time.Time) {
+	next := time.Now().Add(d)
+	s = time.After(d)
+	t = &next
+	return
+}
+
+func (m *gcPolicyManager) run(ctx context.Context) {
+	var (
+		schedC <-chan time.Time
+
+		lastCollection *time.Time
+		nextCollection *time.Time
+
+		interval    = time.Second
+		gcTime      time.Duration
+		collections int
+
+		dirty     bool
+		mutations int
+	)
+	if m.startupDelay > 0 {
+		schedC, nextCollection = schedule(m.startupDelay)
+	}
+	for {
+		select {
+		case <-schedC:
+			if lastCollection != nil && (m.mutationThreshold == 0 || mutations < m.mutationThreshold) && !dirty {
+				schedC, nextCollection = schedule(interval)
+				continue
+			}
+			break
+		case e := <-m.eventC:
+			if lastCollection != nil && lastCollection.After(e.ts) {
+				continue
+			}
+			if e.dirty {
+				dirty = true
+			}
+			mutations++
+
+			// Check if condition should cause immediate collection
+			if (dirty && m.onDelete) ||
+				(m.mutationThreshold > 0 && mutations >= m.mutationThreshold) ||
+				(m.mutationThreshold == 0 && nextCollection == nil) {
+
+				// Check if not already scheduled before delay threshold
+				if nextCollection == nil || nextCollection.After(time.Now().Add(m.scheduleDelay)) {
+					schedC, nextCollection = schedule(m.scheduleDelay)
+				}
+			}
+
+			continue
+		}
+
+		t1 := time.Now()
+		err := m.c.GarbageCollect(ctx)
```

Comment:
The garbage collect method should probably return the pause time, it occurred to me that this calculation is including the time to acquire the lock. During the acquisition time mutations can occur and therefore should not be considered "paused" during that time.